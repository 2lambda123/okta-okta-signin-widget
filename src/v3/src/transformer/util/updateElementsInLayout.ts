/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type MapFn = (uischema: UISchemaElement) => UISchemaElement;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: UISchemaLayout;
  predicate: PredicateFn;
  callback: CallbackFn;
};

type OptionsOld = {
  layout: UISchemaLayout;
  mapFn: MapFn;
  predicateFn: PredicateFn;
};

export const updateElementsInLayout = (options: OptionsOld) => {
  const fn = (
    layout: UISchemaLayout,
    mapFn: MapFn,
    predicateFn: PredicateFn = () => true,
  ): UISchemaLayout => {
    // eslint-disable-next-line no-param-reassign
    layout.elements = layout.elements.map((element) => {
      const { type } = element;
      if (type === UISchemaLayoutType.STEPPER) {
        // eslint-disable-next-line no-param-reassign
        (element as StepperLayout).elements = (element as StepperLayout).elements
          .map((el) => fn(el, mapFn, predicateFn));
        return element;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        return fn(element as UISchemaLayout, mapFn, predicateFn);
      }

      if (predicateFn(element)) {
        return mapFn(element);
      }

      return element;
    });

    return layout;
  };

  return fn(options.layout, options.mapFn, options.predicateFn);
};

export const traveseLayout = (options: Options) => {
  const fn = (
    layout: UISchemaLayout,
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
        return;
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
}
