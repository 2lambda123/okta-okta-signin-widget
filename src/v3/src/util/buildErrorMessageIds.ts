/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const buildErrorMessageIds = (errors: string[], fieldName: string): string => {
  if (errors.length === 1) {
    return `${fieldName}-error`;
  }
  const ids = errors.map((_: string, index: number) => `${fieldName}-error-${index}`);
  return ids.join(' ');
};
