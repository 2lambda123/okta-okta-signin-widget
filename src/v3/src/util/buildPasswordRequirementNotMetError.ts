/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetError = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
): string | undefined => {
  // TODO: Add ticket here to create new translation key
  const errorMessagePrefix = loc('registration.error.password.passwordRequirementsNotMet.prefix', 'login');
  const errorMessageSuffix = passwordRequirements.map((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    return ruleValue === false ? requirement.label : undefined;
  }).filter(Boolean).join(', ') || undefined;

  return typeof errorMessageSuffix !== 'undefined'
    ? `${errorMessagePrefix} ${errorMessageSuffix}`
    : undefined;
};
