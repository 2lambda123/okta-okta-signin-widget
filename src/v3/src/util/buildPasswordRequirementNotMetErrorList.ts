/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessageWithName, ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetErrorList = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
  fieldName: string,
): IdxMessageWithName[] => {
  const errorMessages: IdxMessageWithName[] = [];

  passwordRequirements.forEach((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    if (ruleValue === false) {
      errorMessages.push({
        name: fieldName,
        class: 'ERROR',
        message: requirement.label,
        i18n: { key: '' },
      });
    }
  });

  if (errorMessages.length > 0) {
    // TODO: Add ticket here to create new translation key that includes colon
    errorMessages.unshift({
      name: fieldName,
      class: 'ERROR',
      message: loc('registration.error.password.passwordRequirementsNotMet', 'login'),
      i18n: { key: 'registration.error.password.passwordRequirementsNotMet' },
    });
  }

  return errorMessages;
};
