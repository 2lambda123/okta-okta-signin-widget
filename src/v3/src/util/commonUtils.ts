/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

// eslint-disable-next-line
export const getDisplayName = (WrappedComponent: any): string => (
  WrappedComponent.displayName || WrappedComponent.name || 'Component'
);

export const handleFormFieldChange = (
  setDirtyFn: ((isDirty: boolean) => void) | undefined,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleChange: (path: string, value: any) => void,
  path: string,
  // eslint-disable-next-line
  value: any,
): void => {
  setDirtyFn?.(true);
  handleChange(path, value);
};
