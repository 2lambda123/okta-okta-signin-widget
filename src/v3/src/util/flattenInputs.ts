/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';

export const flattenInputs = (input: Input, name = ''): Input[] => {
  const res: Input[] = [];
  const { value } = input;

  if (Array.isArray(value)) {
    return value.reduce((acc, curr) => [
      ...acc,
      ...flattenInputs(curr, input.name),
    ], res);
  }

  return [{
    ...input,
    name: name ? `${name}.${input.name}` : input.name,
  }];
};
