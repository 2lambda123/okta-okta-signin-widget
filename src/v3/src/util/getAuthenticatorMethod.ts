/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

export const getAuthenticatorMethod = (
  transaction: IdxTransaction,
): string | undefined => {
  // currentAuthenticator is from enrollment flows and currentAuthenticatorEnrollment is from verify flows
  // @ts-ignore Add currentAuthenticatorEnrollment to IdxContext type
  const { context: { currentAuthenticator, currentAuthenticatorEnrollment } } = transaction;
  const enrollMethods = currentAuthenticator?.value?.methods;
  const verifyMethods = currentAuthenticatorEnrollment?.value?.methods;

  return (enrollMethods && enrollMethods[0].type) || (verifyMethods && verifyMethods[0].type);
};
