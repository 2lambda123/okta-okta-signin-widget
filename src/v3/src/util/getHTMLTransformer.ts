/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { convertNodeToElement, Options } from 'react-html-parser';

export const getHTMLTransformer = (
  currentNodeName: string,
  newNodeName: string,
  attributes?: Record<string, unknown>,
): Options['transform'] => {
  const transform: Options['transform'] = (node, index) => {
    if (node.type === 'tag' && node.name === currentNodeName) {
      // eslint-disable-next-line no-param-reassign
      node.name = newNodeName;
      if (typeof attributes !== 'undefined') {
        // eslint-disable-next-line no-param-reassign
        node.attribs = attributes;
      }
      return convertNodeToElement(node, index, transform!);
    }
    return undefined;
  };

  return transform;
};
