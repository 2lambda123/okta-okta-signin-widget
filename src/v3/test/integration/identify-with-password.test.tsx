/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/introspect/default.json';

describe('identify-with-password', () => {
  it('renders the loading state first', async () => {
    const { container, findByLabelText } = await setup({ mockResponse });
    await findByLabelText('Loading...');
    expect(container).toMatchSnapshot();
  });

  it('renders form', async () => {
    const { container, findByLabelText } = await setup({ mockResponse });
    await findByLabelText(/Username/);
    expect(container).toMatchSnapshot();
  });

  describe('sends correct payload', () => {
    it('with all required fields', async () => {
      const { authClient, user, findByTestId } = await setup({ mockResponse });

      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByTestId('submit');

      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        'POST',
        'https://oie-8425965.oktapreview.com/idp/idx/identify',
        {
          data: JSON.stringify({
            identifier: 'testuser@okta.com',
            credentials: {
              passcode: 'fake-password',
            },
            rememberMe: false,
            stateHandle: 'fake-stateHandle',
          }),
          headers: {
            Accept: 'application/json; okta-version=1.0.0',
            'Content-Type': 'application/json',
            'X-Okta-User-Agent-Extended': 'okta-auth-js/9.9.9',
          },
          withCredentials: true,
        },
      );
    });

    it('with required fields + optional fields', async () => {
      const { authClient, user, findByTestId } = await setup({ mockResponse });

      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const rememberMeEl = await findByTestId('rememberMe');
      const submitButton = await findByTestId('submit');

      await user.type(usernameEl, 'testuser@okta.com');
      expect(usernameEl.value).toEqual('testuser@okta.com');
      await user.type(passwordEl, 'fake-password');
      expect(passwordEl.value).toEqual('fake-password');
      await user.click(rememberMeEl);
      await user.click(submitButton);

      expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
        'POST',
        'https://oie-8425965.oktapreview.com/idp/idx/identify',
        {
          data: JSON.stringify({
            identifier: 'testuser@okta.com',
            credentials: {
              passcode: 'fake-password',
            },
            rememberMe: true,
            stateHandle: 'fake-stateHandle',
          }),
          headers: {
            Accept: 'application/json; okta-version=1.0.0',
            'Content-Type': 'application/json',
            'X-Okta-User-Agent-Extended': 'okta-auth-js/9.9.9',
          },
          withCredentials: true,
        },
      );
    });

    it('fails client side validation with no inputs', async () => {
      const {
        authClient,
        user,
        findByTestId,
      } = await setup({ mockResponse });
      let identifierError; 
      let passwordError;

      const usernameEl = await findByTestId('identifier') as HTMLInputElement;
      const passwordEl = await findByTestId('credentials.passcode') as HTMLInputElement;
      const submitButton = await findByTestId('submit');
      await findByTestId('rememberMe');

      // empty username & empty password
      await user.click(submitButton);
      identifierError = await findByTestId('identifier-error');
      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      passwordError = await findByTestId('credentials.passcode-error');
      expect(passwordError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalledWith(
        'POST',
        'https://oie-8425965.oktapreview.com/idp/idx/identify',
        expect.anything(),
      );

      // only username - resubmit can clear errors
      await user.type(usernameEl, 'testuser@okta.com');
      await user.click(submitButton);
      passwordError = await findByTestId('credentials.passcode-error');
      expect(passwordError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalledWith(
        'POST',
        'https://oie-8425965.oktapreview.com/idp/idx/identify',
        expect.anything(),
      );

      // only password - reset clears error
      await user.type(passwordEl, 'fake-password');
      await user.click(submitButton);
      identifierError = await findByTestId('identifier-error');
      expect(identifierError.textContent).toEqual('This field cannot be left blank');
      expect(authClient.options.httpRequestClient).not.toHaveBeenCalledWith(
        'POST',
        'https://oie-8425965.oktapreview.com/idp/idx/identify',
        expect.anything(),
      );
    });
  });
});
