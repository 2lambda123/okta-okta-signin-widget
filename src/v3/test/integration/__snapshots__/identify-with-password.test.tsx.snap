// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identify-with-password renders form 1`] = `
<div>
  <span>
    <main
      class="ods-4o5zYQ ods-76eppy ods-66Tc65 ods-6lar65 ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs auth-container main-container mainViewContainer"
    >
      <div
        class="ods-4o5zYQ ods-2Se4oq ods-2VFzyK ods-1qUaxm ods-1oujX1 ods-78Q1Pv ods-3x5mmc ods-7mcS6o ods-7lpPkQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs siwContainer"
      >
        <div
          class="okta-sign-in-header auth-header siwHeader"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 emotion-0"
          />
        </div>
        <div
          class="MuiBox-root emotion-1"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiBox-root emotion-2"
            >
              <div
                class="MuiBox-root emotion-3"
              >
                <div
                  class="MuiBox-root emotion-4"
                >
                  <h2
                    class="MuiTypography-root MuiTypography-h3 emotion-5"
                  >
                    Sign In
                  </h2>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <div
                  class="MuiBox-root emotion-7"
                >
                  <label
                    class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-8"
                    for="identifier"
                  >
                    Username
                  </label>
                  <div
                    class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth emotion-9"
                  >
                    <input
                      aria-invalid="false"
                      autocomplete="username"
                      class="MuiOutlinedInput-input MuiInputBase-input emotion-10"
                      data-se="identifier"
                      id="identifier"
                      name="identifier"
                      type="text"
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline emotion-11"
                    >
                      <legend
                        class="emotion-12"
                      >
                        <span
                          class="notranslate"
                        >
                          ​
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                  
                </div>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <div
                  class="MuiBox-root emotion-7"
                >
                  <label
                    class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-8"
                    for="credentials.passcode"
                  >
                    Password
                  </label>
                  <div
                    class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedEnd emotion-16"
                    helpertext=""
                  >
                    <input
                      aria-invalid="false"
                      autocomplete="current-password"
                      class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd emotion-17"
                      data-se="credentials.passcode"
                      id="credentials.passcode"
                      name="credentials.passcode"
                      type="password"
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd emotion-18"
                    >
                      <button
                        aria-label="toggle password visibility"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium emotion-19"
                        data-mui-internal-clone-element="true"
                        tabindex="0"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-20"
                          data-testid="VisibilityIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline emotion-11"
                    >
                      <legend
                        class="emotion-12"
                      >
                        <span
                          class="notranslate"
                        >
                          ​
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <label
                  class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd emotion-24"
                >
                  <span
                    class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root emotion-25"
                  >
                    <input
                      class="PrivateSwitchBase-input emotion-26"
                      data-se="rememberMe"
                      id="rememberMe"
                      name="rememberMe"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-20"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label emotion-28"
                  >
                    Remember this device
                  </span>
                </label>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <button
                  class="MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root  emotion-30"
                  data-se="#/properties/submit"
                  tabindex="0"
                  type="submit"
                >
                  Sign in
                </button>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <button
                  class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-32"
                  data-se="forgot-password"
                  tabindex="0"
                  type="button"
                >
                  Forgot password?
                </button>
              </div>
              <div
                class="MuiBox-root emotion-3"
              >
                <button
                  class="MuiButton-root MuiButton-floating MuiButton-floatingPrimary MuiButton-sizeMedium MuiButton-floatingSizeMedium MuiButton-disableElevation MuiButtonBase-root  emotion-32"
                  data-se="enroll"
                  tabindex="0"
                  type="button"
                >
                  Register
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </main>
  </span>
</div>
`;
